contracts/src/chainlink/solidity/emergency/StarkNetValidator.sol

```diff
@@ -6,44 +6,82 @@ import "@chainlink/contracts/src/v0.8/interfaces/TypeAndVersionInterface.sol";
 contract StarkNetValidator is TypeAndVersionInterface, AggregatorValidatorInterface, SimpleWriteAccessController {
+  // Config for L1 -> L2 message cost approximation
+  struct GasConfig {
+    uint256 gasUsed;
+    uint256 buffer;
+    address gasPriceL1Feed;
+  }
+
   int256 private constant ANSWER_SEQ_OFFLINE = 1;
-  // Selector hardcoded because StarkNet hash function is not available in this environment
-  uint256 constant STARK_SELECTOR_UPDATE_STATUS =
-    1585322027166395525705364165097050997465692350398750944680096081848180365267;
 
-  IStarknetMessaging public immutable STARKNET_CROSS_DOMAIN_MESSENGER;
+  uint256 public immutable SELECTOR_STARK_UPDATE_STATUS = _selectorStarkNet("update_status");
   uint256 public immutable L2_UPTIME_FEED_ADDR;
 
-  /// @notice StarkNet messaging contract address - the address is 0.
-  error InvalidStarkNetMessaging();
+  IStarknetMessaging public immutable STARKNET_CROSS_DOMAIN_MESSENGER;
+
+  AggregatorV3Interface private s_source;
+  AccessControllerInterface private s_configAC;
+  GasConfig private s_gasConfig;
+
+  /// @notice This event is emitted when the gas config is set.
+  event GasConfigSet(uint256 gasUsed, uint256 buffer, address indexed gasPriceL1Feed);
+  /// @notice emitted when a new gas access-control contract is set
+  event ConfigACSet(address indexed previous, address indexed current);
+  /// @notice emitted when a new source aggregator contract is set
+  event SourceAggregatorSet(address indexed previous, address indexed current);
 
+  /// @notice StarkNet messaging contract address - the address is 0.
+  error InvalidStarkNetMessagingAddress();
   /// @notice StarkNet uptime feed address - the address is 0.
-  error InvalidUptimeFeedAddress();
+  error InvalidL2FeedAddress();
+  /// @notice Error thrown when the source aggregator address is 0
+  error InvalidSourceAggregatorAddress();
+  /// @notice Error thrown when the l1 gas price feed address is 0
+  error InvalidGasPriceL1FeedAddress();
+  /// @notice Error thrown when caller is not the owner and does not have access
+  error AccessForbidden();
 
   /**
-   * @param starkNetMessaging the address of the StarkNet Messaging contract address
-   * @param l2UptimeFeedAddr the address of the Sequencer Uptime Feed on L2
+   * @param starkNetMessaging the address of the StarkNet Messaging contract
+   * @param configAC the address of the AccessController contract managing config access
+   * @param gasPriceL1Feed address of the L1 gas price feed (used to approximate bridge L1 -> L2 message cost)
+   * @param source the source aggregator that we'll read data from (on retries)
+   * @param l2Feed the address of the target L2 contract (Sequencer Uptime Feed)
    */
-  constructor(address starkNetMessaging, uint256 l2UptimeFeedAddr) {
+  constructor(
+    address starkNetMessaging,
+    address configAC,
+    address gasPriceL1Feed,
+    address source,
+    uint256 l2Feed,
+    uint256 gasUsed,
+    uint256 buffer
```

Comment:
I'd simplify this and just use a single value `s_gasConfig.gasEstimate`, which can include the buffer so no need for an extra config:

```suggestion
    uint256 gasEstimate,
```